### PR TITLE
refactor: DBTP-1596 Separate out the platform-config.yml schema validation (part 1)

### PR DESCRIPTION
**Pull request by WillGibson**
Partially addresses [DBTP-1596](https://uktrade.atlassian.net/browse/DBTP-1596).

This is a first pass at separating the `platform-config.yml` schema validation from other validation concerns.

In readiness for the next effort, which will likely involve converting it to a class, this attempts to:

* organise it more coherently
* more clearly name things
* identify "private/public" elements
* reduce duplication
* remove unused elements

---
### Title:
- [x] Scope included as per [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Ticket reference included (unless it's a quick out of ticket thing)
### Description:
- [x] Link to ticket included (unless it's a quick out of ticket thing)
- [x] Includes tests (or an explanation for why it doesn't)
- ~[ ] If the work includes user interface changes, before and after screenshots included in description~
- ~[ ] Includes any applicable changes to the documentation in this code base~
- ~[ ] Includes link(s) to any applicable changes to the documentation in the [DBT Platform Documentation](https://platform.readme.trade.gov.uk/) (can be to a pull request)~
### Tasks:
- [x] [Run the end to end tests for this branch]([https://github.com/uktrade/platform-tools?tab=readme-ov-file#regression-tests](https://github.com/uktrade/platform-end-to-end-tests?tab=readme-ov-file#running-the-tests)) and confirm that they are passing

Passing run: https://eu-west-2.console.aws.amazon.com/codesuite/codepipeline/pipelines/pull-request-end-to-end-tests/executions/90c68d28-753e-4f1a-98e8-b9db00b8351a/variables?region=eu-west-2

[DBTP-1596]: https://uktrade.atlassian.net/browse/DBTP-1596?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dbt_platform_helper/constants.py [hunk 1]
```diff
@@ -1,6 +1,10 @@
+# Todo: Move to Config provider
 PLATFORM_CONFIG_FILE = "platform-config.yml"
+# Todo: Can we get rid of this yet?
 PLATFORM_HELPER_VERSION_FILE = ".platform-helper-version"
```
Comment (WillGibson) [inline dbt_platform_helper/constants.py:4]: @antroy-madetech ?

File: dbt_platform_helper/utils/platform_config_schema.py [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+import re
+
+from schema import Optional
+from schema import Or
+from schema import Regex
+from schema import Schema
+from schema import SchemaError
+
+
+def _string_matching_regex(regex_pattern: str):
```
Comment (ksugden) [inline dbt_platform_helper/utils/platform_config_schema.py:11]: from the error message, this function is a copilot specific validation so should probably be named as such

File: dbt_platform_helper/constants.py (hunk 1 above)
Comment (antroy-madetech) [reply dbt_platform_helper/constants.py:4]: Yes, I think so. Any teams still using it will be forced to upgrade, so it'll be a breaking change.

Comment (WillGibson) [reply dbt_platform_helper/constants.py:4]: I'm not clear from your answer whether you think we should keep support for this file?

File: dbt_platform_helper/utils/platform_config_schema.py [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+
+
+def _integer_between(lower, upper):
+    def _is_between(value):
+        if isinstance(value, int) and lower <= value <= upper:
+            return True
+        # Todo: Raise suitable PlatformException?
+        raise SchemaError(f"should be an integer between {lower} and {upper}")
+
+    return _is_between
```
Comment (ksugden) [inline dbt_platform_helper/utils/platform_config_schema.py:30]: I wonder if these validation functions would be easier to understand at first pass if they always returned something called `validation_function` or something like that?  I found the idea of `_is_between(value)` a bit confusing.

File: dbt_platform_helper/utils/platform_config_schema.py (hunk 2 above)
Comment (WillGibson) [reply dbt_platform_helper/utils/platform_config_schema.py:11]: It's actually only used to validate the branch name in the codebase pipeline, the bit of the error message was misleading. I have removed it.

File: dbt_platform_helper/utils/platform_config_schema.py (hunk 3 above)
Comment (WillGibson) [reply dbt_platform_helper/utils/platform_config_schema.py]: Reworked in person.

**Review by ksugden — APPROVED** [review]

**Pull request merged**